### ISSUE TITLE
Improve the type error message for unassignable form groups

### ISSUE THREAD

**Issue by mwaibel-go** [angular/angular#56864]
### Which @angular/* package(s) are relevant/related to the feature request?

forms

### Description

Search terms: _FormGroup not assignable_, _TS2322_, _2322_

When a typed reactive form requires data from a service, it cannot be created in the constructor. Rather, we need to declare the form in the component and separately instantiate it after the constructor has run, usually in the ngOnInit hook.

Separating the declaration steps from the initialization step can lead to the two types misaligning, which causes an error like the following:

```
Type 'FormGroup<{ a: FormControl<string>; b: FormControl<string>; c: FormControl<string | null>; }>' is not assignable to type 'FormGroup<{ a: FormControl<string>; b: FormControl<string>; c: FormControl<string>; }>'.
  Types of property 'controls' are incompatible.
    Type '{ a: FormControl<string>; b: FormControl<string>; c: FormControl<string | null>; }' is not assignable to type '{ a: FormControl<string>; b: FormControl<string>; c: FormControl<string>; }'.(2322)
```

[Stackblitz](https://stackblitz.com/edit/stackblitz-starters-4ld5bn?file=src%2Fmain.ts)

This is a small example, but already it gets hard to parse. The issue is that one of the form controls wasn’t marked as nonNullable during initialization, and `FormControl<string|null>` is not assignable to `FormControl<string>`. Now imagine you have a Form with multiple levels of nested FormArrays and FormGroups and one control is out of sync with its declaration. It is a nightmare to debug.

### Proposed solution

When a FormGroup instance is not assignable to a variable of FormType, the error message should indicate the paths of the offending control(s) in both the instantiated and declared types. This could be done similarly to how TypeScript handles incompatible structured types:

```typescript
type Declaration = {
    a: string
    b: {
        b1: string
    }
}

type Instantiation = {
    a: string
    b: {
        b1: string | null
    }
}

let decl: Declaration
let inst!: Instantiation

decl = inst
```

```
Type 'Instantiation' is not assignable to type 'Declaration'.
  The types of 'b.b1' are incompatible between these types.
    Type 'string | null' is not assignable to type 'string'.
      Type 'null' is not assignable to type 'string'.(2322)
```
[TS Playground](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAIhDGAbAhgJ2cAlgewHZQF4oBvAWACgoqpkAuKAZ2FU1wHMLqoAjesyrtW4BGekxbtO1AL4VZ5CqEhQAkribJcWDDnxF+XOo2asOAqrxJTBPUcYlsoAHyi4ArokTWo8+RUQQwFAAJgiI9HBIaDp4-oFQrEwAhPRqGlqYMbgUFKFIhAnqwEA)

I say “in both the instantiated and declared types” to catch the case where one type has more form controls than the other – this case is handled by this error message in TS:

```
Property 'b' is missing in type 'Instantiation' but required in type 'Declaration'.(2741)
```

When two form controls differ in type, then the error message should unwrap the types of the controls and explain why they’re not assignable. In the example above, this might look like this:

```
Type 'FormGroup<{ a: FormControl<string>; b: FormControl<string>; c: FormControl<string | null>; }>' is not assignable to type 'FormGroup<{ a: FormControl<string>; b: FormControl<string>; c: FormControl<string>; }>'.
  Types of property 'controls' are incompatible.
    Type '{ a: FormControl<string>; b: FormControl<string>; c: FormControl<string | null>; }' is not assignable to type '{ a: FormControl<string>; b: FormControl<string>; c: FormControl<string>; }'.
       The types of 'c' are incompatible between these types.
         Type 'string | null' is not assignable to type 'string'.
           Type 'null' is not assignable to type 'string'.(2322)
```

### Alternatives considered

None.

**Comment by JeanMeche** [comment]
Angular is not responsible in for the error, this is purely a typescript issue. 
The difference with your repro is that `FormGroup` uses a generic for its properties, so this isn't actionnable to us.
